### DISCUSSION TITLE
Is it possible to conditionally expand a relationship using OData?

### DISCUSSION THREAD

**Discussion by julealgon** [OData/AspNetCoreOData#458] [Q&A]
Say I have a `QueryCategory` entity which has many `Query` entities (1-to-many classic parent-child relationship).

Now, I want to get all category information, but I only want to expand the queries for a specific category. This would be a sample response:
```json
[{
        "name": "Category1",
        "queries": []
    }, {
        "name": "Category2",
        "queries": [{
                "id": 1,
                "queryName": "Some Query",
                "category": "Category2"
            }, {
                "id": 2,
                "queryName": "Another Query",
                "category": "Category2"
            }
        ]
    }, {
        "name": "Category3",
        "queries": []
    }
]
```
Notice how categories 1 and 3 are present, but their `queries` property is not expanded (its empty). Category 2 has `queries` expanded.

Is this a supported scenario in the OData protocol and .NET implementation using a single request? Can one expand a property on just _some_ of the returned entities?

**Comment by julealgon** [comment]
I read a bit more on this after asking the question and found out about the `$it` alias. I was not able to make it work on the TripPin service. According to the spec, this should work:

http://docs.oasis-open.org/odata/odata/v4.0/odata-v4.0-part2-url-conventions.html

> Example 84: customers along with their orders that shipped to the same city as the customer's address. The nested filter expression is evaluated in the context of Orders; $it allows referring to values in the outer context of Customers.
> 
> `http://host/service/Customers?
            $expand=Orders($filter=$it/Address/City eq ShipTo/City)`

I tried:
- `https://services.odata.org/TripPinRESTierService/People?$expand=Trips($filter=$it/UserName eq 'russellwhyte')`

But it returns an error:
```json
{
  "error": {
    "code": "",
    "message": "Could not find a property named 'UserName' on type 'Trippin.Trip'."
  }
}
```

I've not yet tried locally, will update this once I do.

**Reply by xuzhg** [reply]
I noticed the scenario. And I also did something in 8 version. But, I am not 100% sure whether It can work or not in the latest 8. x version. 

see comments at: https://github.com/OData/AspNetCoreOData/blob/main/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L389-L392

**Reply by julealgon** [reply]
Do you happen to know why it doesn't work on the reference TripPin service @xuzhg ? What server library/version is being used there?
